### PR TITLE
Add hor2eq function

### PR DESCRIPTION
**Pull request by TestSubjector**
The function was a bit tricky to do (the `Real` type keyword arguments, if placed before the `Bool` type keyword arguments lead to type-instability).

Additionally, this is an important function because of the many functions it is dependent upon (good chance of this function's tests breaking if some other function is modified in the future).

### CODE REVIEW

File: src/hor2eq.jl [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+julia> adstring(ra_o, dec_o)
+" 00 13 17.3  +15 11 26"
+```
+
+### Notes ###
+
+Code of this function is based on IDL Astronomy User's Library.
+"""
+hor2eq(alt::Real, az::Real, jd::Real; ws::Bool=false, B1950::Bool=false,
```
Comment (giordano) [inline src/hor2eq.jl]: In `_hor2eq`, put all floating point arguments one after the other, so that you can promote all of them together:
```julia
_hor2eq(promote(float(alt), float(az), float(jd), float(lat), float(lon), float(altitude),
                float(temperature), float(pressure))..., ws,
        B1950, precession, nutate, aberration, refract,
        obsname)
```

File: src/hor2eq.jl [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+This is a function to calculate equatorial (ra,dec) coordinates from
+horizon (alt,az) coords. It is accurate to about 1 arcsecond or better.
+It performs precession, nutation, aberration, and refraction corrections.
+
+### Arguments ###
+
+* `alt`: altitude of horizon coords, in degrees
+* `az`: azimuth angle measured East from North (unless ws is `true`), in degrees
+* `jd`: julian date
+* `ws` (optional boolean keyword): set this to `true` to get the azimuth measured
```
Comment (giordano) [inline src/hor2eq.jl]: Document the default value

File: src/hor2eq.jl [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    if refract
+        alt_b = co_refract(alt, altitude, pressure, temperature)
+    else
+        alt_b = alt
+    end
+
+    if ws
+        az -= 180
+    end
+    dra1, ddec1, eps, d_psi, _ = co_nutate(jd, 45, 45)
```
Comment (giordano) [inline src/hor2eq.jl]: `_` is useful to discard leading or intermediate elements of an iterable (`_, b, _, d = 1, 2, 3, 4`), but for trailing elements there is no need to use `_`,
```julia
_, _, eps, d_psi = co_nutate(jd, 45, 45)
```
should be sufficient ;-)

File: src/hor2eq.jl [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    last = 15 * ct2lst(lon, jd) + d_psi * cos(eps) / 3600
+    ha, dec = altaz2hadec(alt_b, az, lat)
+    ra = mod(last - ha, 360)
+    ha /= 15
+    dra1, ddec1, eps, _, _ = co_nutate(jd, ra, dec)
```
Comment (giordano) [inline src/hor2eq.jl]: Like above,
```julia
dra1, ddec1, eps = co_nutate(jd, ra, dec)
```

File: src/hor2eq.jl [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        if B1950
+            ra, dec = precess(ra, dec, j_now, 1950, FK4=true)
+        else
+            ra, dec = precess(ra, dec, j_now, 2000)
+        end
+    end
+    return ra, dec, ha
+end
+"""
+
```
Comment (giordano) [inline src/hor2eq.jl]: Something missing here?  Please, leave a blank line before the docstring

File: test/utils-tests.jl [hunk 6]
```diff
@@ -346,6 +346,30 @@ end
...
 
+# Test hor2eq
+# The values used for the testset are from running the code. However they have been
+# correlated with the output from hor2eq routine of IDL AstroLib, with
+# differences only in the least significant digits.
+@testset "hor2eq" begin
+    ra_o, dec_o, ha_o = hor2eq(43.6879, 56.684, AstroLib.J2000, B1950=true)
+    @test ra_o ≈ 259.20005705918317
+    @test dec_o ≈ 49.674706171288655
+    @test ha_o ≈ 19.40545272279752
```
Comment (giordano) [inline test/utils-tests.jl]:
```
GDL> hor2eq, 43.6879d0, 56.684d0, 2451545d0, ra_o, dec_o, ha_o, /B1950
GDL> print, ra_o, dec_o, ha_o
       259.20239       49.674907       291.08179
```
Do you have a clue about why `ra_o` and `dec_o` differ already on the 4th-5th most significant digit?

`ha_o` is completely different: you converted to hours, but docstring says it's in degrees, but when I convert back to degrees the value of this quantity exactly matches the value in IDL AstroLib, so there is no problem with it, only the RA and Dec coordinates.

File: src/hor2eq.jl [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    end
+
+    if ws
+        az -= 180
+    end
+    dra1, ddec1, eps, d_psi, _ = co_nutate(jd, 45, 45)
+    last = 15 * ct2lst(lon, jd) + d_psi * cos(eps) / 3600
+    ha, dec = altaz2hadec(alt_b, az, lat)
+    ra = mod(last - ha, 360)
+    ha /= 15
```
Comment (giordano) [inline src/hor2eq.jl]: I don't think `ha` is returned in degrees, not hours, this instruction should be removed (see comment about the result of the test below)

File: src/hor2eq.jl [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+* `aberration` (optional boolean keyword): set this to `false` for no aberration
+  correction, `true` by default
+* `refract` (optional boolean keyword: set this to `false` for no refraction
+  correction, `true` by default
+* `lat` (optional keyword): north geodetic latitude of location, in degrees. Default
+  is NaN
+* `lon` (optional keyword): AST longitude of location, in degrees. You can specify west
+  longitude with a negative sign. Dafault is NaN.
+* `altitude` (optional keyword): the altitude of the observing location, in meters.
+  It it zero by default
```
Comment (giordano) [inline src/hor2eq.jl]: "It is"

File: src/hor2eq.jl [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+* `pressure` (optional keyword): the pressure at the observing location, in millibars.
+  Default is NaN
+* `temperature` (optional keyword): the temperature at the observing location, in Kelvins.
+  Default is NaN
+* `obsname` (optional keyword): set this to a valid observatory name to
+  be used by the observatory type in [types](@ref), which will return
+  the latitude and longitude to be used by this program. Default is NaN.
```
Comment (giordano) [inline src/hor2eq.jl]: This should be defined better.  The default is the empty string `""`, not `NaN`.  In addition, it should be explained that when it's an empty string, `lat` and `lon` defaults to the coordinates of Pine Bluff Observatory.

File: src/hor2eq.jl [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    if ws
+        az -= 180
+    end
+    dra1, ddec1, eps, d_psi, _ = co_nutate(jd, 45, 45)
+    last = 15 * ct2lst(lon, jd) + d_psi * cos(eps) / 3600
+    ha, dec = altaz2hadec(alt_b, az, lat)
+    ra = mod(last - ha, 360)
+    ha /= 15
+    dra1, ddec1, eps, _, _ = co_nutate(jd, ra, dec)
+    dra2, ddec2 = co_aberration(jd, ra, dec, eps)
```
Comment (giordano) [inline src/hor2eq.jl]: Can this be moved inside the `if aberration`?

File: src/hor2eq.jl [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+    if nutate
+       ra -= dra1
```
Comment (giordano) [inline src/hor2eq.jl]: Shouldn't `dra1` and `ddec1` in following line be divided by 3600?

Comment (TestSubjector) [inline src/hor2eq.jl]: `co_aberration.pro` returned the `ra` and `dec` in arcseconds (that's why the division by 3600 to make it into degrees), however `co_aberration.jl` returns the `ra` and `dec` in degrees itself.

File: test/utils-tests.jl (hunk 6 above)
Comment (TestSubjector) [inline test/utils-tests.jl]: I think the difference stems from the use of the updated values in  `true_obliquity` and `mean_obliquity`; in `co_aberration` and `co_nutate` respectively ( #24 )

**Comment by giordano** [comment]
As you noted, type-stability of this function is a bit tricky because of the keyword arguments.  Keyword arguments in Julia are handy, but also often a pain when you seek performance, because they aren't part of dispatch.

The inner function `_hor2eq` instead is completely inferrable, for example try
```julia
@code_warntype AstroLib._hor2eq(43.6879, 56.684, Float64(AstroLib.J2000), NaN, NaN, 0.0, NaN, NaN, false, false, true, true, true, true, "")

@code_warntype AstroLib._hor2eq(big(43.6879), big(56.684), BigFloat(AstroLib.J2000), big(NaN), big(NaN), big(0.0), big(NaN), big(NaN), false, false, true, true, true, true, "")
```

Since at least the inner function looks stable (which is the most important thing), I'd probably merge the PR now, but please add at the end of the file drop a TODO note saying that `hor2eq` isn't stable because of the keyword arguments.

File: src/hor2eq.jl [hunk 12]
```diff
@@ -85,27 +89,29 @@ It performs precession, nutation, aberration, and refraction corrections.
...
+  longitude to be used by this program. This is `""` (empty string) by default,
+  in which case `lat` and `lon` default to the coordinates of the `Pine Bluff Observatory`
+  provided they are equivalent to `NaN` individually
 
 ### Output ###
 
 * `ra`: right ascension of object, in degrees (FK5)
 * `dec`: declination of the object, in degrees (FK5)
-* `ha`: hour angle, in degrees
+* `ha`: hour angle, in hours
```
Comment (giordano) [inline src/hor2eq.jl]: Why this change?  I'd have kept the unit in degrees, for consistency with IDL AstroLib

File: src/hor2eq.jl [hunk 13]
```diff
@@ -116,12 +122,12 @@ and the local time is 10 PM precisely. What is the right ascension and declinati
...
-julia> ra_o, dec_o = hor2eq(ten(37,54,41), ten(264,55,06), 2466879.7083333, obsname="kpno",
-                            pressure = 711, temperature = 273)
-(3.3222851503189124, 15.190605763758745, 3.640795457403172)
+julia> ra_o, dec_o = hor2eq(ten(37,54,41), ten(264,55,06), 2466879.7083333,
+                            obsname="kpno", pressure = 711, temperature = 273)
+(3.32228485671625, 15.19060567248328, 3.640795457403172)
 
 julia> adstring(ra_o, dec_o)
-" 00 13 17.3  +15 11 26"
+" 00 13 17.3  +15 11 26
```
Comment (giordano) [inline src/hor2eq.jl]: Remove the closing quote by error?

File: src/hor2eq.jl [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+        # Using Pine Bluff Observatory values
+        if isnan(lat)
+            lat = T(43.0783)
+        end
+
+        if isnan(lon)
+            lon = T(-89.865)
+        end
+    else
+        lat = observatories[obsname].latitude
```
Comment (TestSubjector) [inline src/hor2eq.jl]: Thanks. I had done that till f818e8e actually, changed that in attempt for type stability and didn't revert it back later

File: src/hor2eq.jl (hunk 13 above)
Comment (TestSubjector) [inline src/hor2eq.jl]: Yes

File: src/hor2eq.jl (hunk 12 above)
Comment (TestSubjector) [inline src/hor2eq.jl]: I assumed that you wanted to change the docstring to hours, my bad. Updated.

**Comment by TestSubjector** [comment]
Made a small attempt at using @inferred in tests, as it gives the side knowledge that  `_hor2eq` has a inferable return type.
